### DISCUSSION TITLE
Has anyone got the tests to work?

### DISCUSSION THREAD

**Discussion by TimJentzsch** [DioxusLabs/taffy#46] [Q&A] [answered]
When using `cargo run --package gentest` to run the tests, I get the following error:

```
   Compiling gentest v0.1.0 (/home/tim/Documents/stretch/scripts/gentest)
    Finished dev [unoptimized + debuginfo] target(s) in 2.16s
     Running `target/debug/gentest`
Starting ChromeDriver 101.0.4951.41 (93c720db8323b3ec10d056025ab95c23a31997c9-refs/branch-heads/4951@{#904}) on port 4444
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
thread 'tokio-runtime-worker-4' panicked at 'unexpected webdriver error; webdriver returned error: unknown error: DevToolsActivePort file doesn't exist', /home/tim/.cargo/registry/src/github.com-1ecc6299db9ec823/fantoccini-0.11.9/src/session.rs:328:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'tokio-runtime-worker-3' panicked at 'called `Option::unwrap()` on a `None` value', /home/tim/.cargo/registry/src/github.com-1ecc6299db9ec823/fantoccini-0.11.9/src/session.rs:282:68
```

Did anyone get this to work and might be able to help me out?

**Comment by TimJentzsch** [comment]
Docs for reference: <https://github.com/DioxusLabs/stretch#testing>

**Comment by alice-i-cecile** [comment]
For future readers, this is discussed further in #45.

**Reply by TimJentzsch** [reply]
This is actually a different issue, #45 is about installing `chromedriver`, which is already completed in this one (see the `Starting ChromeDriver`... line in the log).

I wasn't able to resolve this particular issue yet.

**Comment by CobaltCause** [comment]
I've hit a different error while trying to use this tool:

```
$ cargo run -p gentest
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
     Running `target/debug/gentest`
Starting ChromeDriver 101.0.4951.41 (93c720db8323b3ec10d056025ab95c23a31997c9-refs/branch-heads/4951@{#904}) on port 4444
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 1/238
thread 'tokio-runtime-worker-1' panicked at 'called `Option::unwrap()` on a `None` value', scripts/gentest/src/main.rs:173:61
```

Extra things I needed to install were `chromedriver`, `chromium`, `pkg-config`, and `openssl`; and then I was able to build the package but running it produced the above error.

**Comment by NickAcPT** [comment, marked as answer]
This isn't a direct answer to this - although I've tried generating the tests but couldn't also get it to work.

While working on my Kotlin port of this same library, I decided to automatically generate the tests myself, based on those same Yoga test files, but while doing so, I came across some inconsistencies with the results given by Sprawl and the ones given by a Chrome browser.

For full transparency, here's the Chrome version I am using: 
> Chrome is up to date
> Version 102.0.5005.63 (Official Build) (64-bit)

A test that shows me these kinds of inconsistencies is [the min_height test](https://github.com/DioxusLabs/sprawl/blob/78c5e05ef0549f2594717c672763519162977372/tests/generated/min_height.rs). 

## Comparison of results in min_height

Just to make my point (and suffering) clear, here's my comparison of the assertions vs what Chrome tells me.

### Root element
```rust
assert_eq!(sprawl.layout(node).unwrap().size.width, 100f32);
assert_eq!(sprawl.layout(node).unwrap().size.height, 100f32);
assert_eq!(sprawl.layout(node).unwrap().location.x, 0f32);
assert_eq!(sprawl.layout(node).unwrap().location.y, 0f32);
```
![Root is 100x100 on Chrome](https://i.imgur.com/QnlLscc.png)
Clearly the root is 100x100 and at 0x0, this checks out.

### Root's first child (node 0)
```rust
assert_eq!(sprawl.layout(node0).unwrap().size.width, 100f32);
assert_eq!(sprawl.layout(node0).unwrap().size.height, 80f32);
assert_eq!(sprawl.layout(node0).unwrap().location.x, 0f32);
assert_eq!(sprawl.layout(node0).unwrap().location.y, 0f32);
```
![Node 0 is 100x60 and at 0x0](https://i.imgur.com/kdyNHKV.png)
This is.. odd? Sprawl is giving the wrong value here, since the test passes, when compared to what Chrome returns.

## Conclusion

I think it's that time of the year to fix the test generation and regenerate those tests and make sure we're still up-to-spec.

###### Please someone help, I've spent 3 hours yesterday trying to debug this issue.

**Reply by alice-i-cecile** [reply]
> I think it's that time of the year to fix the test generation and regenerate those tests and make sure we're still up-to-spec.

Agreed. I don't *love* using Chrome as a source of truth, but it's not a terrible plan.

**Reply by NickAcPT** [reply]
> > I think it's that time of the year to fix the test generation and regenerate those tests and make sure we're still up-to-spec.
> 
> Agreed. I don't _love_ using Chrome as a source of truth, but it's not a terrible plan.

Well, apparently that's already being done from what I can see. It just appears to be out-of-date.

**Reply by alice-i-cecile** [reply]
Yes; it's just that this is an inherited library (from the dead stretch crate). There are quite a few design choices that I feel... differently about.

**Reply by NickAcPT** [reply]
> Yes; it's just that this is an inherited library (from the dead stretch crate). There are quite a few design choices that I feel... differently about.

Well, fair. Now it's just a matter of whether or not it's worth the effort to change some of those design choices and refactor stuff.

**Reply by alice-i-cecile** [reply]
Yep. Next course of action will be to add unit tests, for the refactored layout algorithm in #47, and then sit down and decide on our overall test strategy.

**Comment by Weibye** [comment]
Update on this: 

@mockersf has helpfully fixed the gentest now in #145, and I have improved the error message in #147 

As a policy: Whenever making changes / refactors that affects these, please run the gentest at the end then commit them separately so that the PR can be reviewed by commit without all the noise.
